### ISSUE TITLE
split_incoming_synapse removes a synapse without pruning memetic — structural_add_neuron silently dies on fine-tuned creatures (structural.rs:999)

### ISSUE THREAD

**Issue by nleck** [stSoftwareAU/NEAT-AI-Lamarck#197]
## Summary

Lamarck removes a synapse without pruning the creature's `memetic` record. The rule is: **any pass that removes a neuron or a synapse must delete or prune `memetic`** — the record holds per-neuron bias deltas and per-synapse weight deltas keyed to a specific structure, so removing structure leaves those keys dangling.

`grep -rn memetic lamarck/src` returns **zero hits**. Lamarck has no concept of the field at all: it never sets it, prunes it, or reads it. Unlike the creature-level `uuid` (which `CreatureExport` has no field for, so it is dropped by construction), `CreatureExport` **does** carry `memetic: Option<MemeticExport>`, and `creature_to_json` re-emits it — so Lamarck carries an inherited memetic straight through every clone and every write.

## The exact site

`lamarck/src/structural.rs:999`, in `split_incoming_synapse`:

```rust
let old = creature.synapses[syn_idx].clone();
...
let old_w = old.weight;
creature.synapses.remove(syn_idx);        // <-- removal, memetic untouched
let bridged = add_neuron_bridge(creature, NeuronBridgeSpec { ... });
if bridged.is_err() {
    creature.synapses.insert(syn_idx, old);
}
```

The creature being mutated is a clone of the incumbent that inherited `memetic: Some(...)` — `lamarck/src/candidates.rs:958`:

```rust
let mut creature = incumbent.clone();
```

`incumbent` comes from `load_creature(&original_text)` → `parse_creature_json`, which populates `memetic`. Reached for `CandidateStrategy::StructuralAddNeuron` at `lamarck/src/candidates.rs:1140-1141`.

Two other `retain`/`remove` calls on `.neurons` / `.synapses` exist — `structural.rs:936` and `structural.rs:967-970` — but both are rollbacks of structure created moments earlier, keyed on the brand-new uuid. Net-zero, no dangling keys possible. `structural.rs:999` is the only real removal.

## The consequence today: a whole strategy silently disappears

This does not currently corrupt a file, and the reason is worth stating precisely because it is accidental rather than designed.

`split_incoming_synapse` delegates to `add_neuron_bridge`, which certifies the result at `structural.rs:962-972` via `crate::validate::validate_creature` → `neat_core::creature_validate` with `LAMARCK_VALIDATE_OPTIONS`. That runs **rule 31 (MEMETIC)**, which enforces *referential integrity*, not just shape — `neat-core/src/creature_validate.rs:1695-1700`:

```rust
if !pairs.contains(&(from_index, to_index)) {
    return Err(ValidationFailure::validation(
        reason::MEMETIC,
        format!("Memetic from id {from_uuid} to id {to_uuid} has no matching synapses."),
    ));
}
```

`pairs` is built from the creature's *live* synapse endpoints. So a memetic weight naming the synapse we just removed makes the bridge fail. `structural.rs:1012-1014` then restores the edge, and `candidates.rs:1141` swallows the error with `.ok()?`.

Net effect: **on any creature that carries a memetic weight row touching the chosen edge, `structural_add_neuron` silently stops producing candidates.** No log line, no counter, no journal entry — the strategy just quietly yields nothing, and the run reports fewer candidates without ever saying why. Since memetic is exactly what a Backprop or fine-tuning stage leaves behind, this bites hardest on the creatures that have had the most work done to them.

## The residual risk: nothing on the write path validates

The protection above lives in one mutation helper, not at the boundary. **No write path validates.** `write_candidate_batch` (`candidates.rs:1240-1266`), the `tags.rs` serialisers, and every `best.json` write emit `memetic` unchecked, and `neat_core::creature_to_json` enforces only `validate_creature_width` — rule 31 is opt-in and is **not** wired into serialisation. `combos::merge_candidate_deltas` (`combos.rs:211-215`, `let mut out = base.clone();`) never calls `validate_creature` at all.

So the moment any removal is added that does not route through `add_neuron_bridge`, Lamarck writes a creature with a dangling memetic straight to disk. That failure surfaces far downstream — this wire format has already cost a full day of Backprop and Lamarck outages with `Creature JSON error: invalid type: sequence, expected a map`. A dangling reference is the same class of problem with the same latency to detection.

## Scope note

A tags-only change is **not** structural: such a pass must leave both `uuid` and `memetic` untouched. Per-neuron `uuid` is a separate concept — a stable identity label and an input to the creature hash — and must be preserved, never cleared.

Pure-append paths are **not** in scope for this rule and should not be "fixed": `structural.rs:842` / `:857` (`insert_synapse_ordered`, `add_synapse`), `grafts.rs:348-436` (`apply_graft`), `combos.rs:248` / `:295`. Adding structure leaves existing memetic keys referentially valid. Lamarck also never assigns `NeuronExport::id` in non-test code (grown neurons get `id: None`), so id-keyed memetic biases cannot dangle through renumbering either.

## TDD — write these first, watch them fail

1. **Red first — the strategy loss.** Build a creature carrying a `memetic` record whose `weights` name an incoming synapse of the focus neuron. Call `split_incoming_synapse` on exactly that edge. Today it returns `Err` (rule 31) and the candidate vanishes. Assert instead that it **succeeds**, and that the resulting creature's `memetic` no longer references the removed edge:
   ```rust
   assert!(result.is_ok());
   let m = creature.memetic.as_ref().unwrap();
   // no weight row / entry naming the removed (from, to) pair
   ```

2. **Red first — the write boundary.** Drive a full `StructuralAddNeuron` acceptance on a memetic-carrying creature and assert that `neat_core::creature_validate(&written, &LAMARCK_VALIDATE_OPTIONS)` passes on the bytes actually written to `best.json` and to `candidates-exp-N/baseline.json`. This is the test that would catch any *future* removal path, not just this one.

3. **The over-correction guard (must stay green).** A pass that changes **only tags** must leave both `memetic` and `uuid` byte-identical to the input. Equally: a pure-append edit (`add_synapse`, `apply_graft`) must **keep** the existing memetic — pruning there would throw away valid fine-tuning history for no reason.

4. **Preserve `MemeticExport::extra`.** `neat-core/src/creature.rs:191-192` flattens unknown memetic keys into `extra`, documented as preserving `generation` / `score` / `ancestry`. Assert those survive a prune. This is what rules out the blunt `memetic = None` fix.

## Suggested fix — prune, structurally, and not here

Do **not** scatter `memetic = None` across the removal sites, for two reasons: it destroys `MemeticExport::extra` (see test 4), and Lamarck is not the only repo that will need this.

Preferred: add the prune to `neat-core` as the single home of the rule — something like `MemeticExport::prune_to(&CreatureExport)` (drop biases whose key no longer resolves, drop weight rows/entries whose `(from, to)` pair is gone, keep `extra` intact) plus a `CreatureExport::prune_memetic()` wrapper. `neat-core` currently exposes **no** prune or clear helper at all (`grep 'clear_memetic|prune_memetic'` → 0 hits), which is why every downstream repo would otherwise hand-roll the blunt version. `neat-core`'s `if_graft.rs` already owns the "single home of that construction rule" idiom, so this fits its existing structure.

Then in Lamarck: call `prune_memetic()` immediately after `creature.synapses.remove(syn_idx)` at `structural.rs:999`, before `add_neuron_bridge` certifies. The rollback branch (`structural.rs:1012-1014`) restores from a clone, so snapshot the memetic alongside `old` and restore it on the error path.

Finally, consider asserting the invariant at the boundary rather than relying on one helper: run `validate_creature` in `write_candidate_batch` and in the `best.json` writers, so a dangling memetic becomes impossible to *write*, not merely impossible to *construct via this one function*.

Related, and worth a separate look: `combos::merge_candidate_deltas` can only express additions and value changes — it has no branch that removes a synapse the base has. A member variant produced by `split_incoming_synapse` had an edge removed; the merged creature silently keeps the base's edge alongside the new bridge. Out of scope for this issue, but it is in the same code family.

**Comment by stservice** [comment]
✅ **Vibe Coder released this claim** — host `vibe-coder-33145`, finished 02:02 UTC.
**Attempts on this issue:** 2
- 01:59 `vibe-coder-13083` — raised #200
- 02:02 `vibe-coder-33145` — released

**Comment by stservice** [comment]
🔀 **Dependency PR opened:** https://github.com/stSoftwareAU/NEAT-AI-core/pull/576

The root cause was fixed in `stSoftwareAU/NEAT-AI-core` on branch `lamarck-197-memetic-prune`, and the worker opened the PR there on the agent's behalf (the agent's own `gh` boundary covers this repo only). Releasing that dependency stays a human decision — this run will not merge or publish it.

**Comment by stservice** [comment]
Pull request https://github.com/stSoftwareAU/NEAT-AI-Lamarck/pull/200 has been created to address this issue.

**Comment by stservice** [comment]
## Issue run model stats

- **Requested model:** `opus`
- **Served model(s):** `claude-opus-5`
- **Effort:** `high`
- **Issue invocations:** 1
- **Tokens:** input 246 · output 93,634 · cache write 234,112 · cache read 22,661,542
- **Prompt cache:** 99.0% (read 22,661,542 · write 234,112 · uncached 246)
- **Turns:** 138
- **Duration:** 25m 14s
- **Estimated cost (USD, estimate only):** ~$15.14
  - `claude-opus-5`: $15.14 — input $0.0012 · output $2.34 · cache write $1.46 · cache read $11.33
- **Degraded:** no

_Estimate only — these figures cover the run(s) this Vibe Coder made on this issue. Earlier phases, and any other Vibe Coder that worked the same issue, are not included._
